File: src/cdisc_library_api/specs/cdash_product.clj
(ns cdisc-library-api.specs.cdash-product
  (:require [clojure.spec.alpha :as s]
            [spec-tools.data-spec :as ds]
            [cdisc-library-api.specs.cdash-product-links :refer :all]
            [cdisc-library-api.specs.cdash-class :refer :all]
            [cdisc-library-api.specs.cdash-domain :refer :all]
            )
  (:import (java.io File)))


(def cdash-product-data
  {
   (ds/opt :name) string?
   (ds/opt :label) string?
   (ds/opt :description) string?
   (ds/opt :source) string?
   (ds/opt :effectiveDate) string?
   (ds/opt :registrationStatus) string?
   (ds/opt :version) string?
   (ds/opt :_links) cdash-product-links-spec
   (ds/opt :classes) (s/coll-of cdash-class-spec)
   (ds/opt :domains) (s/coll-of cdash-domain-spec)
   })

(def cdash-product-spec
  (ds/spec
    {:name ::cdash-product
     :spec cdash-product-data}))

File: client/scalaz/src/main/scala/org/openapitools/client/api/CdashProduct.scala
package org.openapitools.client.api

import argonaut._
import argonaut.EncodeJson._
import argonaut.DecodeJson._

import org.http4s.{EntityDecoder, EntityEncoder}
import org.http4s.argonaut._
import org.joda.time.DateTime


import CdashProduct._

case class CdashProduct (
  name: Option[String],
label: Option[String],
description: Option[String],
source: Option[String],
effectiveDate: Option[String],
registrationStatus: Option[String],
version: Option[String],
links: Option[CdashProductLinks],
classes: Option[List[CdashClass]],
domains: Option[List[CdashDomain]])

object CdashProduct {
  import DateTimeCodecs._

  implicit val CdashProductCodecJson: CodecJson[CdashProduct] = CodecJson.derive[CdashProduct]
  implicit val CdashProductDecoder: EntityDecoder[CdashProduct] = jsonOf[CdashProduct]
  implicit val CdashProductEncoder: EntityEncoder[CdashProduct] = jsonEncoderOf[CdashProduct]
}
